File: Java/src/session2/td/heritageMultiple/ZeroParInt.java
package session2.td.heritageMultiple;

import session1.td.FabriqueNaturels;
import session1.td.Nat;

public class ZeroParInt extends EtatZero implements AlgebreNatParInt {

    public static final FabriqueNaturels<Nat> FAB = new ZeroParInt();

    public ZeroParInt() {
    }

    @Override
    public Nat creerZero() {
        return new ZeroParInt();
    }

    @Override
    public Nat creerSuccesseur(Nat predecesseur) {
        return new SuccDecimal(predecesseur);
    }

    @Override
    public String toString() {
        return ""+this.val();
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof Nat))
            return false;
        Nat x = (Nat)obj;
        return this.val() == x.val();
    }
}

File: TypeScript/src/session2.td/heritageMultiple/ZeroParInt.ts
import {Nat} from "../../session1.td/Nat";
import {FabriqueNaturels} from "../../session1.td/FabriqueNaturels";
import {EtatZero} from "./EtatZero";
import {AlgebreNatParInt, AlgebreNatParInt_default} from "./AlgebreNatParInt";
import {SuccDecimal} from "./SuccDecimal";
import {heriterInterfacesConcretes} from "./heritageInterfacesconcretes";

export class ZeroParInt extends EtatZero implements AlgebreNatParInt {

    public static FAB : FabriqueNaturels<Nat> = new ZeroParInt();

    public creerZero() : Nat {
        return new ZeroParInt();
    }

    public creerSuccesseur(predecesseur) {
        return new SuccDecimal(predecesseur);
    }

    public toString() : string {
        return String(this.val());
    }

    public equals(obj : Nat) : boolean {
        return this.val() == obj.val();
    }

    div(x: Nat): Nat {
        return undefined;
    }

    modulo(x: Nat): Nat {
        return undefined;
    }

    produit(x: Nat): Nat {
        return undefined;
    }

    somme(x: Nat): Nat {
        return undefined;
    }

    un(): Nat {
        return undefined;
    }

    zero(): Nat {
        return undefined;
    }
}

heriterInterfacesConcretes(ZeroParInt, [EtatZero, AlgebreNatParInt_default]);